sklearn/preprocessing/_encoders.py

```diff
@@ -791,9 +811,17 @@ def transform(self, X):
         X_out : sparse matrix or a 2-d array
             Transformed input.
         """
-        X_int, X_mask = self._transform(X, handle_unknown=self.handle_unknown)
+        X_int, X_mask = self._transform(X, handle_unknown=self.handle_unknown,
+                                        force_all_finite='allow-nan')
         X_trans = X_int.astype(self.dtype, copy=False)
 
+        for cat_idx, categories_for_idx in enumerate(self.categories_):
+            missing_indices = [i for i, v in enumerate(categories_for_idx)
+                               if is_scalar_nan(v) or v is None]
+            for missing_idx in missing_indices:
```

Comment:
ordinarily would we expect `len(missing_indices) <= 1`. I look at this code and immediately incline to reduce its complexity, but if `len(missing_indices)` is rarely more than 1 then it's fine.